File: src/secantus/sortkey.py
"""Byte-sortable encoding of BSON values for index keys.

The encoder produces ``bytes`` such that lexicographic byte comparison matches
MongoDB's BSON cross-type sort order (so numbers sort below strings, strings
below objects, etc., and within a type values sort by their natural order).
That lets us put encoded keys in WiredTiger's sorted B-tree and use range
scans for ``$gt`` / ``$gte`` / ``$lt`` / ``$lte`` / ``$in`` and for
sort-by-indexed-field, not just equality.

The encoding is one-way: it preserves order and equality, not the original
value (so we don't decode it back). The original document is still the BSON
blob in ``table:secantus_documents``; the index entry's ``key_bytes`` is just
a sortable fingerprint.

Layout: ``<rank_byte><payload>``. ``rank_byte`` is the BSON type rank
(MinKey=1 .. MaxKey=13). Payload format depends on type — see ``_encode_*``
helpers. Compound keys are joined with ``\\x00\\x00`` after payload nulls
have been escaped to ``\\x00\\xff``, so the join is unambiguous and
byte-sortable.

Numeric encoding (the hard one) uses a Decimal-based "lexical decimal" form
so int / long / double / Decimal128 collide on equal value and sort
correctly across the whole numeric type. Specials (NaN, ±Infinity) get
dedicated bytes that bracket the finite range.
"""

from __future__ import annotations

import datetime as _dt
import math
import struct
from collections.abc import Mapping
from decimal import Decimal, InvalidOperation
from typing import Any

import bson
from bson import Binary, Decimal128, MaxKey, MinKey, ObjectId, Regex, Timestamp

# Type ranks — must match storage._bson_type_rank.
RANK_MINKEY = 1
RANK_NULL = 2
RANK_NUMBER = 3
RANK_STRING = 4
RANK_DOCUMENT = 5
RANK_ARRAY = 6
RANK_BINDATA = 7
RANK_OBJECTID = 8
RANK_BOOL = 9
RANK_DATE = 10
RANK_TIMESTAMP = 11
RANK_REGEX = 12
RANK_MAXKEY = 13


def _rank(value: Any) -> int:
    if isinstance(value, MinKey):
        return RANK_MINKEY
    if value is None:
        return RANK_NULL
    if isinstance(value, bool):
        return RANK_BOOL
    if isinstance(value, (int, float, Decimal128)):
        return RANK_NUMBER
    if isinstance(value, str):
        return RANK_STRING
    if isinstance(value, Mapping):
        return RANK_DOCUMENT
    if isinstance(value, list):
        return RANK_ARRAY
    if isinstance(value, (bytes, Binary, bytearray)):
        return RANK_BINDATA
    if isinstance(value, ObjectId):
        return RANK_OBJECTID
    if isinstance(value, _dt.datetime):
        return RANK_DATE
    if isinstance(value, Timestamp):
        return RANK_TIMESTAMP
    if isinstance(value, Regex):
        return RANK_REGEX
    if isinstance(value, MaxKey):
        return RANK_MAXKEY
    return RANK_DOCUMENT  # unknown -> treat as object


# Byte-sortable decimal markers within the number payload.
_NUM_NAN = b"\x00"
_NUM_NEG_INF = b"\x20"
_NUM_NEG = 0x40  # prefix byte for finite negative
_NUM_ZERO = b"\x80"
_NUM_POS = 0xC0  # prefix byte for finite positive
_NUM_POS_INF = b"\xff"


def _to_decimal(value: int | float | Decimal128) -> Decimal | None:
    """Return value as a Decimal, or None for NaN/Infinity."""
    if isinstance(value, Decimal128):
        try:
            d = value.to_decimal()
        except (InvalidOperation, ValueError):
            return None
        if not d.is_finite():
            return None
        return d
    if isinstance(value, float):
        if math.isnan(value) or math.isinf(value):
            return None
        try:
            return Decimal(repr(value))
        except (InvalidOperation, ValueError):
            return None
    return Decimal(int(value))


def _is_nan(value: int | float | Decimal128) -> bool:
    if isinstance(value, float):
        return math.isnan(value)
    if isinstance(value, Decimal128):
        try:
            return value.to_decimal().is_nan()
        except (InvalidOperation, ValueError):
            return False
    return False


def _inf_sign(value: int | float | Decimal128) -> int:
    """Return +1 for +inf, -1 for -inf, 0 if not infinity."""
    if isinstance(value, float):
        if value == math.inf:
            return 1
        if value == -math.inf:
            return -1
        return 0
    if isinstance(value, Decimal128):
        try:
            d = value.to_decimal()
        except (InvalidOperation, ValueError):
            return 0
        if d.is_infinite():
            return -1 if d.is_signed() else 1
    return 0


def _encode_number(value: int | float | Decimal128) -> bytes:
    if _is_nan(value):
        return _NUM_NAN
    inf = _inf_sign(value)
    if inf > 0:
        return _NUM_POS_INF
    if inf < 0:
        return _NUM_NEG_INF
    d = _to_decimal(value)
    if d is None or d.is_zero():
        return _NUM_ZERO
    sign = 1 if d > 0 else -1
    d = abs(d).normalize()
    digits, exp = d.as_tuple()[1:3]
    sci_exp = exp + len(digits) - 1
    bias_e = 128 + sci_exp
    if not 0 <= bias_e <= 255:
        # Out of single-byte exponent range — fall back to a marker that
        # still sorts on the correct side of zero. Magnitudes within ranks
        # collapse together; acceptable until we widen the encoding.
        return bytes([_NUM_POS, 0xFF, 0xFF]) if sign > 0 else bytes([_NUM_NEG, 0x00, 0x00])
    if sign < 0:
        bias_e = 0xFF - bias_e
    if len(digits) % 2 == 1:
        digits = digits + (0,)
    pairs = bytes(da * 10 + db + 1 for da, db in zip(digits[::2], digits[1::2], strict=True))
    if sign < 0:
        pairs = bytes(0x64 - p for p in pairs)  # invert: 100 - encoded
    prefix = _NUM_POS if sign > 0 else _NUM_NEG
    terminator = b"\x00" if sign > 0 else b"\xff"
    return bytes([prefix, bias_e]) + pairs + terminator


def _escape(data: bytes) -> bytes:
    """Replace 0x00 bytes so 0x00 0x00 can be a compound separator."""
    return data.replace(b"\x00", b"\x00\xff")


def _encode_string(s: str, collation: Any = None) -> bytes:
    """UTF-8 byte-sortable encoding for a string, optionally
    collation-normalised.

    When ``collation`` is set and supports index encoding, the string
    is normalised (accents stripped / case-folded per the collation's
    ``strength`` and ``case_level``) before encoding so the entries
    table's lex byte order matches the collation's sort order. Index
    entries are written with the SAME collation that queries use to
    look them up, so two strings that compare-equal under the
    collation produce the same key bytes and hit the same row.

    ``numericOrdering`` is intentionally not supported at the index
    level (would need a length-prefixed digit-run encoding for
    sortability); queries that combine ``numericOrdering`` with an
    index fall back to COLLSCAN per the picker's contract.
    """
    if collation is not None and getattr(collation, "supports_index_encoding", False):
        # Local import — sortkey.py is on the import-cycle floor;
        # collation.py imports sortkey via storage, so we can't
        # top-import.
        from secantus.collation import normalize_for_index_bytes

        return _escape(normalize_for_index_bytes(s, collation))
    return _escape(s.encode("utf-8"))


def _encode_binary(b: bytes | Binary | bytearray) -> bytes:
    raw = bytes(b)
    return struct.pack(">I", len(raw)) + _escape(raw)


def _encode_objectid(oid: ObjectId) -> bytes:
    return oid.binary  # 12 bytes, already byte-sortable


def _encode_date(d: _dt.datetime) -> bytes:
    if d.tzinfo is not None:
        ms = int(d.timestamp() * 1000)
    else:
        epoch = _dt.datetime(1970, 1, 1)
        ms = int((d - epoch).total_seconds() * 1000)
    return _signed_int64_sortable(ms)


def _signed_int64_sortable(n: int) -> bytes:
    # Map signed int64 to unsigned by flipping the top bit, then pack big-endian.
    # Negatives map below zero (top-bit-cleared), positives map above (top-bit-set).
    n &= 0xFFFFFFFFFFFFFFFF  # truncate just in case
    return struct.pack(">Q", n ^ 0x8000000000000000)


def _encode_timestamp(ts: Timestamp) -> bytes:
    return struct.pack(">II", ts.time, ts.inc)


def _encode_doc(doc: Mapping[str, Any]) -> bytes:
    return _escape(bson.encode(dict(doc)))


def _encode_array(arr: list[Any]) -> bytes:
    # Encode as a BSON document with positional keys — this matches how
    # BSON itself stores arrays, so equality through the index lines up
    # with equality at the matches() layer.
    return _escape(bson.encode({str(i): v for i, v in enumerate(arr)}))


def _encode_regex(r: Regex) -> bytes:
    pattern = r.pattern.encode("utf-8") if isinstance(r.pattern, str) else bytes(r.pattern)
    flags = r.flags.encode("utf-8") if isinstance(r.flags, str) else bytes(r.flags)
    return _escape(pattern) + b"\x00\x00" + _escape(flags)


def encode_value(value: Any, *, collation: Any = None) -> bytes:
    """Single-value byte-sortable BSON encoding.

    When ``collation`` is set and the value is a string, the string
    is normalised via :func:`secantus.collation.normalize_for_index_bytes`
    before encoding so the resulting bytes sort by the collation's
    rules rather than raw codepoint. Non-string values pass through
    unchanged. Used by index-write and index-lookup paths to keep
    string entries collation-aware.
    """
    rank = _rank(value)
    head = bytes([rank])
    if rank in (RANK_MINKEY, RANK_NULL, RANK_MAXKEY):
        return head
    if rank == RANK_NUMBER:
        return head + _encode_number(value)
    if rank == RANK_STRING:
        return head + _encode_string(value, collation)
    if rank == RANK_DOCUMENT:
        return head + _encode_doc(value)
    if rank == RANK_ARRAY:
        return head + _encode_array(value)
    if rank == RANK_BINDATA:
        return head + _encode_binary(value)
    if rank == RANK_OBJECTID:
        return head + _encode_objectid(value)
    if rank == RANK_BOOL:
        return head + (b"\x01" if value else b"\x00")
    if rank == RANK_DATE:
        return head + _encode_date(value)
    if rank == RANK_TIMESTAMP:
        return head + _encode_timestamp(value)
    if rank == RANK_REGEX:
        return head + _encode_regex(value)
    return head  # unreachable


COMPOUND_SEP = b"\x00\x00"


def encode_compound(values: list[Any], *, collation: Any = None) -> bytes:
    """Compound key. Components are null-escaped; ``\\x00\\x00`` separates.

    ``collation`` applies uniformly to every string component (matches
    mongod — a collation is a per-index property, not per-field).
    """
    return COMPOUND_SEP.join(encode_value(v, collation=collation) for v in values)


def invert_bytes(b: bytes) -> bytes:
    """Bitwise-NOT every byte. Order-reversing: if a < b byte-wise, then ~a > ~b.

    Used to store descending-direction index entries so a forward B-tree
    walk yields values in descending order without per-row reversal.
    """
    return bytes(x ^ 0xFF for x in b)


def encode_value_directed(value: Any, direction: int = 1, *, collation: Any = None) -> bytes:
    """Like ``encode_value`` but inverts bytes when ``direction == -1``."""
    e = encode_value(value, collation=collation)
    return invert_bytes(e) if direction == -1 else e


# Range-query bound helpers. Bounds are returned as (key_bytes, inclusive)
# tuples so the WT range scan can apply them with the right boundary
# semantics. ``None`` for a bound means open-ended.
def gt_bound(value: Any, *, collation: Any = None) -> tuple[bytes, bool]:
    return encode_value(value, collation=collation), False


def gte_bound(value: Any, *, collation: Any = None) -> tuple[bytes, bool]:
    return encode_value(value, collation=collation), True


def lt_bound(value: Any, *, collation: Any = None) -> tuple[bytes, bool]:
    return encode_value(value, collation=collation), False


def lte_bound(value: Any, *, collation: Any = None) -> tuple[bytes, bool]:
    return encode_value(value, collation=collation), True

File: crates/secantus-core/src/sortkey.rs
//! Byte-sortable BSON value encoding — Rust port of `secantus.sortkey`.
//!
//! Produces bytes whose lexicographic order matches MongoDB's BSON cross-type
//! sort order, so index entries can live in WiredTiger's sorted B-tree. This
//! is the first leaf engine ported under the Python -> Rust rewrite
//! (tasks/rust-rewrite-plan.md Phase 1); it is the byte-exact counterpart of
//! the pure-Python `encode_value` and is pinned against it by a parity test
//! (`tests/test_rust_sortkey_parity.py`) and the cargo unit tests below.
//!
//! Collation-normalised string keys and the single-byte-exponent overflow
//! fallback are reproduced; `numericOrdering` index encoding is intentionally
//! out of scope (matches Python — those queries fall back to COLLSCAN).

use bson::{Bson, Document};

use crate::collation::{self, Collation};

// Type ranks — must match secantus.sortkey.
const RANK_MINKEY: u8 = 1;
const RANK_NULL: u8 = 2;
const RANK_NUMBER: u8 = 3;
const RANK_STRING: u8 = 4;
const RANK_DOCUMENT: u8 = 5;
const RANK_ARRAY: u8 = 6;
const RANK_BINDATA: u8 = 7;
const RANK_OBJECTID: u8 = 8;
const RANK_BOOL: u8 = 9;
const RANK_DATE: u8 = 10;
const RANK_TIMESTAMP: u8 = 11;
const RANK_REGEX: u8 = 12;
const RANK_MAXKEY: u8 = 13;

const NUM_NAN: u8 = 0x00;
const NUM_NEG_INF: u8 = 0x20;
const NUM_NEG: u8 = 0x40;
const NUM_ZERO: u8 = 0x80;
const NUM_POS: u8 = 0xC0;
const NUM_POS_INF: u8 = 0xFF;

/// Compound-key separator (also the escape sentinel). Matches
/// `secantus.sortkey.COMPOUND_SEP`.
pub const COMPOUND_SEP: &[u8] = b"\x00\x00";

/// Error type for values the encoder doesn't handle (Python's `encode_value`
/// doesn't handle them either — Symbol/Code/etc. route to its document branch
/// and raise). The Python shim treats this as "fall back to the pure-Python
/// path", so we never silently diverge.
#[derive(Debug)]
pub struct UnsupportedValue(pub String);

impl std::fmt::Display for UnsupportedValue {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "sortkey: unsupported BSON value: {}", self.0)
    }
}

/// 0x00 -> 0x00 0xff, order-preserving so 0x00 0x00 is an unambiguous separator.
fn escape(data: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(data.len());
    for &b in data {
        out.push(b);
        if b == 0 {
            out.push(0xff);
        }
    }
    out
}

/// Decimal as value = sign * (digits) * 10^exp, no leading/trailing zero
/// digits. `None` for zero / non-finite.
struct DecimalParts {
    sign: i32,
    digits: Vec<u8>,
    exp: i64,
}

fn normalize(digits: &mut Vec<u8>, exp: &mut i64) {
    while digits.first() == Some(&0) {
        digits.remove(0);
    }
    while digits.last() == Some(&0) {
        digits.pop();
        *exp += 1;
    }
}

fn parse_decimal_str(s: &str) -> Option<DecimalParts> {
    let s = s.trim();
    let (sign, rest) = match s.strip_prefix('-') {
        Some(r) => (-1, r),
        None => (1, s.strip_prefix('+').unwrap_or(s)),
    };
    let (mantissa, exp_extra) = match rest.split_once(['e', 'E']) {
        Some((m, e)) => (m, e.parse::<i64>().ok()?),
        None => (rest, 0),
    };
    let (int_part, frac_part) = match mantissa.split_once('.') {
        Some((i, f)) => (i, f),
        None => (mantissa, ""),
    };
    let mut digits: Vec<u8> = Vec::new();
    for c in int_part.chars().chain(frac_part.chars()) {
        digits.push(c.to_digit(10)? as u8);
    }
    let mut exp = exp_extra - frac_part.len() as i64;
    normalize(&mut digits, &mut exp);
    if digits.is_empty() {
        return None; // zero
    }
    Some(DecimalParts { sign, digits, exp })
}

fn parts_from_int(mut n: i128) -> Option<DecimalParts> {
    if n == 0 {
        return None;
    }
    let sign = if n < 0 { -1 } else { 1 };
    if n < 0 {
        n = -n;
    }
    let mut digits: Vec<u8> = n.to_string().bytes().map(|b| b - b'0').collect();
    let mut exp: i64 = 0;
    normalize(&mut digits, &mut exp);
    Some(DecimalParts { sign, digits, exp })
}

fn encode_number_from_parts(p: &DecimalParts) -> Vec<u8> {
    let sci_exp = p.exp + p.digits.len() as i64 - 1;
    let mut bias_e = 128 + sci_exp;
    if !(0..=255).contains(&bias_e) {
        // Out of single-byte exponent range — sort on the correct side of
        // zero, magnitudes within the rank collapse (matches Python).
        return if p.sign > 0 {
            vec![NUM_POS, 0xFF, 0xFF]
        } else {
            vec![NUM_NEG, 0x00, 0x00]
        };
    }
    if p.sign < 0 {
        bias_e = 0xFF - bias_e;
    }
    let mut digits = p.digits.clone();
    if digits.len() % 2 == 1 {
        digits.push(0);
    }
    let mut pairs: Vec<u8> = digits
        .chunks_exact(2)
        .map(|c| c[0] * 10 + c[1] + 1)
        .collect();
    if p.sign < 0 {
        for b in pairs.iter_mut() {
            *b = 0x64 - *b;
        }
    }
    let prefix = if p.sign > 0 { NUM_POS } else { NUM_NEG };
    let terminator = if p.sign > 0 { 0x00 } else { 0xff };
    let mut out = vec![prefix, bias_e as u8];
    out.append(&mut pairs);
    out.push(terminator);
    out
}

fn encode_number(b: &Bson) -> Vec<u8> {
    let parts = match b {
        Bson::Int32(n) => parts_from_int(*n as i128),
        Bson::Int64(n) => parts_from_int(*n as i128),
        Bson::Double(d) => {
            if d.is_nan() {
                return vec![NUM_NAN];
            }
            if d.is_infinite() {
                return vec![if *d > 0.0 { NUM_POS_INF } else { NUM_NEG_INF }];
            }
            if *d == 0.0 {
                None
            } else {
                // Python: Decimal(repr(value)); Rust's shortest Display is the
                // equivalent round-tripping decimal for a finite f64.
                parse_decimal_str(&format!("{d}"))
            }
        }
        Bson::Decimal128(d) => {
            let s = d.to_string();
            let low = s.to_lowercase();
            if low.contains("nan") {
                return vec![NUM_NAN];
            }
            if low.contains("inf") {
                return vec![if s.starts_with('-') {
                    NUM_NEG_INF
                } else {
                    NUM_POS_INF
                }];
            }
            parse_decimal_str(&s)
        }
        _ => unreachable!("encode_number on non-number"),
    };
    match parts {
        None => vec![NUM_ZERO],
        Some(p) => encode_number_from_parts(&p),
    }
}

fn signed_int64_sortable(n: i64) -> [u8; 8] {
    ((n as u64) ^ 0x8000_0000_0000_0000).to_be_bytes()
}

/// Encode an array as a BSON document with positional string keys, matching
/// `secantus.sortkey._encode_array` (so array equality lines up at the index).
fn encode_array_bytes(arr: &[Bson]) -> Result<Vec<u8>, UnsupportedValue> {
    let mut doc = Document::new();
    for (i, v) in arr.iter().enumerate() {
        doc.insert(i.to_string(), v.clone());
    }
    doc_to_escaped_bytes(&doc)
}

fn doc_to_escaped_bytes(doc: &Document) -> Result<Vec<u8>, UnsupportedValue> {
    let mut buf = Vec::new();
    doc.to_writer(&mut buf)
        .map_err(|e| UnsupportedValue(format!("doc encode failed: {e}")))?;
    Ok(escape(&buf))
}

/// Byte-sortable encoding of a single BSON value. Byte-exact counterpart of
/// `secantus.sortkey.encode_value`. `coll` is the index's collation (or `None`);
/// it applies only to top-level string values — strings nested inside documents
/// / arrays are encoded as raw BSON, matching Python's `_encode_doc` /
/// `_encode_array` (which don't thread collation).
pub fn encode_value(v: &Bson, coll: Option<&Collation>) -> Result<Vec<u8>, UnsupportedValue> {
    let mut out = Vec::new();
    match v {
        Bson::MinKey => out.push(RANK_MINKEY),
        Bson::Null => out.push(RANK_NULL),
        Bson::MaxKey => out.push(RANK_MAXKEY),
        Bson::Int32(_) | Bson::Int64(_) | Bson::Double(_) | Bson::Decimal128(_) => {
            out.push(RANK_NUMBER);
            out.extend(encode_number(v));
        }
        Bson::String(s) => {
            out.push(RANK_STRING);
            let bytes = match coll {
                Some(c) => collation::normalize_index_bytes(s, c)
                    .ok_or_else(|| UnsupportedValue("collation defers to Python".into()))?,
                None => s.as_bytes().to_vec(),
            };
            out.extend(escape(&bytes));
        }
        Bson::Document(d) => {
            out.push(RANK_DOCUMENT);
            out.extend(doc_to_escaped_bytes(d)?);
        }
        Bson::Array(a) => {
            out.push(RANK_ARRAY);
            out.extend(encode_array_bytes(a)?);
        }
        Bson::Binary(b) => {
            out.push(RANK_BINDATA);
            out.extend_from_slice(&(b.bytes.len() as u32).to_be_bytes());
            out.extend(escape(&b.bytes));
        }
        Bson::ObjectId(oid) => {
            out.push(RANK_OBJECTID);
            out.extend_from_slice(&oid.bytes());
        }
        Bson::Boolean(b) => {
            out.push(RANK_BOOL);
            out.push(if *b { 1 } else { 0 });
        }
        Bson::DateTime(dt) => {
            out.push(RANK_DATE);
            out.extend_from_slice(&signed_int64_sortable(dt.timestamp_millis()));
        }
        Bson::Timestamp(ts) => {
            out.push(RANK_TIMESTAMP);
            out.extend_from_slice(&ts.time.to_be_bytes());
            out.extend_from_slice(&ts.increment.to_be_bytes());
        }
        Bson::RegularExpression(r) => {
            out.push(RANK_REGEX);
            out.extend(escape(r.pattern.as_bytes()));
            out.extend_from_slice(COMPOUND_SEP);
            out.extend(escape(r.options.as_bytes()));
        }
        other => return Err(UnsupportedValue(format!("{other:?}"))),
    }
    Ok(out)
}

/// Bitwise-NOT every byte — order-reversing, for descending index entries.
pub fn invert_bytes(b: &[u8]) -> Vec<u8> {
    b.iter().map(|x| x ^ 0xFF).collect()
}

/// `encode_value`, bytes inverted when `direction == -1`.
pub fn encode_value_directed(
    v: &Bson,
    direction: i32,
    coll: Option<&Collation>,
) -> Result<Vec<u8>, UnsupportedValue> {
    let e = encode_value(v, coll)?;
    Ok(if direction == -1 { invert_bytes(&e) } else { e })
}

#[cfg(test)]
mod tests {
    use super::*;
    use bson::Bson;

    fn ev(v: Bson) -> Vec<u8> {
        encode_value(&v, None).unwrap()
    }

    #[test]
    fn ranks_order_across_types() {
        // null < number < string < bool < date (sample of the rank ladder).
        assert!(ev(Bson::Null) < ev(Bson::Int32(0)));
        assert!(ev(Bson::Int32(0)) < ev(Bson::String("".into())));
        assert!(ev(Bson::String("z".into())) < ev(Bson::Boolean(false)));
    }

    #[test]
    fn cross_type_numeric_collision() {
        // The headline property: equal numeric value -> identical key bytes,
        // regardless of int32 / int64 / double / decimal128 representation.
        let i = ev(Bson::Int32(3));
        let l = ev(Bson::Int64(3));
        let d = ev(Bson::Double(3.0));
        let dec = ev(Bson::Decimal128("3".parse().unwrap()));
        assert_eq!(i, l);
        assert_eq!(i, d);
        assert_eq!(i, dec);
    }

    #[test]
    fn numbers_sort_correctly() {
        let mut keys = [
            ev(Bson::Double(-2.5)),
            ev(Bson::Int32(-1)),
            ev(Bson::Int32(0)),
            ev(Bson::Double(1.5)),
            ev(Bson::Int32(1000)),
        ];
        let sorted = keys.clone();
        keys.sort();
        assert_eq!(
            keys, sorted,
            "encoded numbers must already be in value order"
        );
    }

    #[test]
    fn directed_inverts_for_descending() {
        let asc = ev(Bson::Int32(5));
        let desc = encode_value_directed(&Bson::Int32(5), -1, None).unwrap();
        assert_eq!(desc, invert_bytes(&asc));
    }
}
